### ISSUE TITLE
relaxed i32x4.trunc_sat_f32x4_{s,u}

### ISSUE THREAD

**Issue by ngzhian** [WebAssembly/relaxed-simd#21]
1. What are the instructions being proposed?

Relaxed versions of `i32x4.trunc_sat_f32x4_s` and `i32x4.trunc_sat_f32x4_u` from Simd128. (Names undecided)

2. What are the semantics of these instructions?

Convert f32x4 to i32x4 with truncation (signed/unsigned). If the inputs are out of range or NaNs, the result is implementation-defined.

3. How will these instructions be implemented? Give examples for at least
   x86-64 and ARM64. Also provide reference implementation in terms of 128-bit
   Wasm SIMD.

## x86/64

relaxed i32x4.trunc_sat_f32x4_s = CVTTPS2DQ
relaxed i32x4.trunc_sat_f32x4_u = VCVTTPS2UDQ (AVX512), Simd128 i32x4.trunc_sat_f32x4_u otherwise (can be slightly optimized to ignore NaNs)

## ARM64

Fcvtzs, Fcvtzu

## ARM

vcvt.S32.F32, vcvt.U32.F32

## simd128

`i32x4.trunc_sat_f32x4_s` and `i32x4.trunc_sat_f32x4_u`

4. How does behavior differ across processors? What new fingerprinting surfaces will be exposed?

Different between x86/64 and ARM/ARM64 for `i32x4.trunc_sat_f32x4_s`:

- x86/64 will return `0x8000000` in lanes for out of range or NaNs
- ARM/ARM64 will return `0`

For `i32x4.trunc_sat_f32x4_u`, it will depend on implementation choice on x86/64:

- if AVX512 is available, same as above, x86/64 will return `0x8000000` in lanes for out of range or NaNs, ARM/ARM64 will return `0`
- if implemented using Simd128, then no difference in the result (but x86/64 requires more instruction counts)

5. What use cases are there?

Conversion instructions are common, if the application can guarantee the input range we can get good performance on all architectures.

**Comment by Maratyszcza** [comment]
IIRC @zeux had a use-case for these instructions.

It would be useful to consider `f64x2` variants in the same proposal.

**Comment by Maratyszcza** [comment]
> For `i32x4.trunc_sat_f32x4_u`, it will depend on implementation choice on x86/64:
> 
> * if AVX512 is available, same as above, x86/64 will return `0x8000000` in lanes for out of range or NaNs, ARM/ARM64 will return `0`

AVX512 version returns 0xFFFFFFFF

**Comment by ngzhian** [comment]
> AVX512 version returns 0xFFFFFFFF
Corrected, thanks!

**Comment by ngzhian** [comment]
> It would be useful to consider f64x2 variants in the same proposal.

relaxed i64x2.trunc_sat_f64x2_{s,u}? We don't have these instructions in Simd128, so I think it is neater to separate them out.

**Comment by Maratyszcza** [comment]
> relaxed i64x2.trunc_sat_f64x2_{s,u}? We don't have these instructions in Simd128, so I think it is neater to separate them out.

The WebAssembly/simd#383 instructions

**Comment by ngzhian** [comment]
i32x4.trunc_sat_f64x2_u_zero and i32x4.trunc_sat_f64x2_s_zero?

**Comment by Maratyszcza** [comment]
Yes

**Comment by zeux** [comment]
Yeah this one is pretty fundamental for many workflows, e.g. in rendering domains it's common to store data as fixed-point integers for GPU consumption but to prepare this data you do some math in floating point and then convert to integer via smth like `int(v * 65535.0f + 0.5f)` (assuming the value is known to be positive); the float->int truncation can be pretty hot based on the amount of other computation.

It would be nice to also include the rounding variants (on x64 assuming default rounding mode setup you can use cvtps2dq for rounding conversion and cvttps2dq for truncating; unsure what floating point environment is typically used in browser context, if it's undefined then rounding would require vroundps before cvttps).

**Comment by yurydelendik** [comment]
What will be the exact recipe for relaxed `i32x4.trunc_sat_f32x4_u` for x86/64 without AVX512? The [comment at #247](https://github.com/WebAssembly/simd/pull/247#issuecomment-641717078) suggests somewhat long version.

Is the following acceptable or the shorter version exists?

-   `y = relaxed i32x4.trunc_sat_f32x4_u(x)` is lowered to:
     *   `MOVAPD xmm_y, xmm_x`
     *   `MOVAPD xmm_tmp, [wasm_i32x4_splat(0x4f000000)]`
     *   `CMPLTPS xmm_tmp, xmm_x`
     *   `PAND xmm_tmp, xmm_x`
     *   `PXOR xmm_y, xmm_tmp`
     *   `CVTTPS2DQ xmm_y, xmm_y`
     *   `PSLLD xmm_tmp, 7`
     *   `PADDD xmm_y, xmm_tmp`

**Comment by ngzhian** [comment]
it will be `CVTTPS2DQ`. The relaxed version only guarantees output when inputs are < INT32_MAX and not NaN, which is exactly what CVTTPS2DQ is, which is available since SSE2.

**Comment by Maratyszcza** [comment]
@ngzhian The question was about the unsigned version, and IIUC we don't expect unsigned version to use just `CVTTPS2DQ` alone.

**Comment by ngzhian** [comment]
Oh oops, sorry I missed that. Hm, then we should reconsider if we want the unsigned version in this. AVX512 is not supported by V8 yet.

**Comment by Maratyszcza** [comment]
IMO it is worth to have unsigned version, both for symmetry and because is it still faster on SSE4.1 than the non-relaxed unsigned version.

**Comment by zeux** [comment]
Should these instructions have `_sat` in the name? In the SIMD MVP _sat stands for saturating, but these instructions don't specify exact behavior for out of range inputs.

**Comment by yurydelendik** [comment]
Agree, there was a mistake 😞  One more operation is needed to make PSLLD work: `PADDD xmm_tmp. xmm_tmp ;  PSLLD xmm_tmp, 8`.

**Comment by ngzhian** [comment]
> Agree, there was a mistake 😞 One more operation is needed to make PSLLD work: ADDPS xmm_tmp. xmm_tmp ; PSLLD xmm_tmp, 8.

Perfect, what a neat trick :) thanks!

**Comment by ngzhian** [comment]
Note: RISC-V V saturates for [same width conversions](https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc#1317-single-width-floating-pointinteger-type-convert-instructions). For f64x2->i32x4 it changes the vector type, and I think there's no guarantee that the top are zeroed.

**Comment by ngzhian** [comment]
On PowerPC VSX xscvdpsxws and xscvdpuxds perform trunc sat

**Comment by ngzhian** [comment]
I think I got the results wrong in this description, ARM/ARM64 doesn't return 0, it saturates.
